weaver/processes/wps_package.py

```diff
@@ -2137,52 +2197,115 @@ def make_location_input(input_type, input_definition):
             if "location" in cwl_result[output_id]:
-                result_loc = cwl_result[output_id]["location"].replace("file://", "")
-                result_wps = os.path.join(wps_out_dir, os.path.split(result_loc)[-1])
-                if os.path.realpath(result_loc) != os.path.realpath(result_wps):
-                    LOGGER.info("Moving [%s]: [%s] -> [%s]", output_id, result_loc, result_wps)
-                    shutil.move(result_loc, result_wps)
-                self.response.outputs[output_id].as_reference = True
-                self.response.outputs[output_id].file = result_wps
-                LOGGER.info("Resolved WPS output [%s]: [%s]", output_id, result_wps)
-            else:
-                result_loc = cwl_result[output_id]
-                result_wps = os.path.join(wps_out_dir, os.path.split(result_loc)[-1])
-                if os.path.realpath(result_loc) != os.path.realpath(result_wps):
-                    LOGGER.info("Moving: [%s] -> [%s]", result_loc, result_wps)
-                    shutil.move(result_loc, result_wps)
-                self.response.outputs[output_id].data = result_wps
-                LOGGER.info("Resolved WPS output [%s]: [%s]", output_id, result_wps)
+                self.make_location_output(cwl_result, output_id)
+                continue
+
+            # data output
+            self.response.outputs[output_id].data = cwl_result[output_id]
+            self.response.outputs[output_id].as_reference = False
+            self.logger.info("Resolved WPS output [%s] as literal data", output_id)
+
+    def make_location_output(self, cwl_result, output_id):
+        # type: (CWLResults, AnyStr) -> None
+        """
+        Rewrite the `WPS` output with required location using result path from `CWL` execution.
+
+        Configures the parameters such that `PyWPS` will either auto-resolve the local paths to match with URL
+        defined by ``weaver.wps_output_url`` or upload it to `S3` bucket from ``weaver.wps_output_s3_bucket`` and
+        provide reference directly.
+
+        .. seealso::
+            - :func:`weaver.wps.load_pywps_config`
+        """
+        wps_out_dir = self.workdir  # pywps will resolve file paths for us using its WPS request UUID
+        s3_bucket = self.settings.get("weaver.wps_output_s3_bucket")
+        result_loc = cwl_result[output_id]["location"].replace("file://", "")
+        result_fn = os.path.split(result_loc)[-1]
+
+        if s3_bucket:
+            # result_wps = "s3://{}/{}".format(s3_bucket, result_fn)
+            # when 'url' is directly enforced, 'ComplexOutput.json' will use it instead of 'file' from temp workdir
+            # self.response.outputs[output_id].url = result_wps
+
+            # override builder only here so that only results are uploaded to S3, and not XML status also
+            # settings are retrieved from PyWPS server config
+            self.response.outputs[output_id]._storage = S3StorageBuilder().build()  # noqa: W0212
```

Comment:
Wow this add-on to pywps was a gift!

Reply:
Yes, and I discovered it only by chance while debug stepping through the fetch call! Very lucky.